File: flutter_custom_tabs_android/android/src/main/java/com/github/droibit/flutter/plugins/customtabs/CustomTabsLauncher.java
package com.github.droibit.flutter.plugins.customtabs;

import static android.content.Intent.FLAG_ACTIVITY_CLEAR_TOP;
import static android.content.Intent.FLAG_ACTIVITY_SINGLE_TOP;
import static androidx.browser.customtabs.CustomTabsService.ACTION_CUSTOM_TABS_CONNECTION;
import static java.util.Objects.requireNonNull;

import android.app.Activity;
import android.app.ActivityManager;
import android.content.ActivityNotFoundException;
import android.content.ComponentName;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.content.pm.ResolveInfo;
import android.net.Uri;
import android.os.Build;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.VisibleForTesting;
import androidx.browser.customtabs.CustomTabsIntent;
import androidx.core.content.ContextCompat;
import androidx.core.util.Pair;

import com.github.droibit.flutter.plugins.customtabs.Messages.FlutterError;
import com.github.droibit.flutter.plugins.customtabs.core.CustomTabsIntentFactory;
import com.github.droibit.flutter.plugins.customtabs.core.ExternalBrowserLauncher;
import com.github.droibit.flutter.plugins.customtabs.core.NativeAppLauncher;
import com.github.droibit.flutter.plugins.customtabs.core.PartialCustomTabsLauncher;
import com.github.droibit.flutter.plugins.customtabs.core.options.CustomTabsIntentOptions;
import com.github.droibit.flutter.plugins.customtabs.core.options.CustomTabsSessionOptions;
import com.github.droibit.flutter.plugins.customtabs.core.session.CustomTabsSessionController;
import com.github.droibit.flutter.plugins.customtabs.core.session.CustomTabsSessionFactory;

import java.util.Map;
import java.util.Objects;

import io.flutter.Log;

public class CustomTabsLauncher implements Messages.CustomTabsApi {
    @VisibleForTesting
    static final String CODE_LAUNCH_ERROR = "LAUNCH_ERROR";

    private final @NonNull CustomTabsIntentFactory customTabsIntentFactory;
    private final @NonNull CustomTabsSessionFactory customTabsSessionFactory;
    private final @NonNull NativeAppLauncher nativeAppLauncher;
    private final @NonNull ExternalBrowserLauncher externalBrowserLauncher;
    private final @NonNull PartialCustomTabsLauncher partialCustomTabsLauncher;
    private @Nullable Activity activity;

    CustomTabsLauncher() {
        this(
                new CustomTabsIntentFactory(),
                new CustomTabsSessionFactory(),
                new NativeAppLauncher(),
                new ExternalBrowserLauncher(),
                new PartialCustomTabsLauncher()
        );
    }

    @VisibleForTesting
    CustomTabsLauncher(
            @NonNull CustomTabsIntentFactory customTabsIntentFactory,
            @NonNull CustomTabsSessionFactory customTabsSessionFactory,
            @NonNull NativeAppLauncher nativeAppLauncher,
            @NonNull ExternalBrowserLauncher externalBrowserLauncher,
            @NonNull PartialCustomTabsLauncher partialCustomTabsLauncher
    ) {
        this.customTabsIntentFactory = customTabsIntentFactory;
        this.customTabsSessionFactory = customTabsSessionFactory;
        this.nativeAppLauncher = nativeAppLauncher;
        this.externalBrowserLauncher = externalBrowserLauncher;
        this.partialCustomTabsLauncher = partialCustomTabsLauncher;
    }

    void setActivity(@Nullable Activity activity) {
        customTabsSessionFactory.handleActivityChange(activity);
        this.activity = activity;
    }

    @Override
    public void launch(
            @NonNull String urlString,
            @NonNull Boolean prefersDeepLink,
            @Nullable Map<String, Object> options
    ) {
        final Activity activity = this.activity;
        if (activity == null) {
            throw new FlutterError(CODE_LAUNCH_ERROR, "Launching a custom tab requires a foreground activity.", null);
        }

        final Uri uri = Uri.parse(urlString);
        if (prefersDeepLink && nativeAppLauncher.launch(activity, uri)) {
            return;
        }

        try {
            final CustomTabsIntentOptions customTabsOptions = customTabsIntentFactory
                    .createIntentOptions(options);
            if (externalBrowserLauncher.launch(activity, uri, customTabsOptions)) {
                return;
            }

            final CustomTabsIntent customTabsIntent = customTabsIntentFactory.createIntent(
                    activity,
                    requireNonNull(customTabsOptions),
                    customTabsSessionFactory
            );
            if (partialCustomTabsLauncher.launch(activity, uri, customTabsIntent)) {
                return;
            }
            customTabsIntent.launchUrl(activity, uri);
        } catch (ActivityNotFoundException e) {
            throw new FlutterError(CODE_LAUNCH_ERROR, e.getMessage(), null);
        }
    }

    @Override
    public void closeAllIfPossible() {
        final Activity activity = this.activity;
        if (activity == null) {
            return;
        }
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.M) {
            return;
        }

        final ActivityManager am = ContextCompat.getSystemService(activity, ActivityManager.class);
        final ComponentName selfActivityName = new ComponentName(activity, activity.getClass());
        //noinspection DataFlowIssue
        for (ActivityManager.AppTask appTask : am.getAppTasks()) {
            final ActivityManager.RecentTaskInfo taskInfo = appTask.getTaskInfo();
            if (!Objects.equals(selfActivityName, taskInfo.baseActivity) ||
                    taskInfo.topActivity == null) {
                continue;
            }
            final Intent serviceIntent = new Intent(ACTION_CUSTOM_TABS_CONNECTION)
                    .setPackage(taskInfo.topActivity.getPackageName());

            if (resolveService(activity.getPackageManager(), serviceIntent, 0) != null) {
                try {
                    final Intent intent = new Intent(activity, activity.getClass())
                            .setFlags(FLAG_ACTIVITY_CLEAR_TOP | FLAG_ACTIVITY_SINGLE_TOP);
                    activity.startActivity(intent);
                } catch (ActivityNotFoundException ignored) {
                }
                break;
            }
        }
    }

    /**
     * @noinspection SameParameterValue
     */
    @SuppressWarnings("deprecation")
    private static @Nullable ResolveInfo resolveService(
            @NonNull PackageManager pm,
            @NonNull Intent intent,
            int flags
    ) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            return pm.resolveService(
                    intent,
                    PackageManager.ResolveInfoFlags.of(flags)
            );
        } else {
            return pm.resolveService(intent, flags);
        }
    }

    @Override
    public @Nullable String warmup(@Nullable Map<String, Object> options) {
        final Activity activity = this.activity;
        if (activity == null) {
            Log.w("CustomTabsAndroid", "Activity is null. Cannot warm up custom tabs.");
            return null;
        }

        final CustomTabsSessionOptions sessionOptions = customTabsSessionFactory
                .createSessionOptions(options);
        final Pair<String, CustomTabsSessionController> session = customTabsSessionFactory
                .createSession(activity, sessionOptions);
        if (session == null) {
            return null;
        }

        final String packageName = session.first;
        final CustomTabsSessionController controller = session.second;
        if (controller.bindCustomTabsService(activity, packageName)) {
            return packageName;
        }
        return null;
    }

    @Override
    public void invalidate(@NonNull String packageName) {
        customTabsSessionFactory.invalidateSession(packageName);
    }
}

File: flutter_custom_tabs_android/android/src/main/kotlin/com/github/droibit/flutter/plugins/customtabs/CustomTabsLauncher.kt
package com.github.droibit.flutter.plugins.customtabs

import android.app.Activity
import android.app.ActivityManager
import android.content.ActivityNotFoundException
import android.content.ComponentName
import android.content.Intent
import android.content.Intent.FLAG_ACTIVITY_CLEAR_TOP
import android.content.Intent.FLAG_ACTIVITY_SINGLE_TOP
import android.content.pm.PackageManager
import android.content.pm.ResolveInfo
import android.os.Build
import androidx.annotation.RestrictTo
import androidx.annotation.VisibleForTesting
import androidx.browser.customtabs.CustomTabsService.ACTION_CUSTOM_TABS_CONNECTION
import androidx.core.content.getSystemService
import androidx.core.net.toUri
import com.github.droibit.flutter.plugins.customtabs.core.CustomTabsIntentFactory
import com.github.droibit.flutter.plugins.customtabs.core.ExternalBrowserLauncher
import com.github.droibit.flutter.plugins.customtabs.core.NativeAppLauncher
import com.github.droibit.flutter.plugins.customtabs.core.PartialCustomTabsLauncher
import com.github.droibit.flutter.plugins.customtabs.core.session.CustomTabsSessionManager

@RestrictTo(RestrictTo.Scope.LIBRARY)
internal class CustomTabsLauncher @VisibleForTesting constructor(
  private val customTabsIntentFactory: CustomTabsIntentFactory,
  private val customTabsSessionManager: CustomTabsSessionManager,
  private val nativeAppLauncher: NativeAppLauncher,
  private val externalBrowserLauncher: ExternalBrowserLauncher,
  private val partialCustomTabsLauncher: PartialCustomTabsLauncher
) : CustomTabsApi {
  private var activity: Activity? = null

  constructor() : this(
    CustomTabsIntentFactory(),
    CustomTabsSessionManager(),
    NativeAppLauncher(),
    ExternalBrowserLauncher(),
    PartialCustomTabsLauncher()
  )

  fun setActivity(activity: Activity?) {
    customTabsSessionManager.handleActivityChange(activity)
    this.activity = activity
  }

  override fun launch(
    urlString: String,
    prefersDeepLink: Boolean,
    options: Map<String, Any>?
  ) {
    val activity = this.activity
      ?: throw FlutterError(
        CODE_LAUNCH_ERROR,
        "Launching a Custom Tab requires a foreground activity.",
        null
      )

    val uri = urlString.toUri()
    if (prefersDeepLink && nativeAppLauncher.launch(activity, uri)) {
      return
    }

    try {
      val customTabsOptions = customTabsIntentFactory.createIntentOptions(options)
      if (externalBrowserLauncher.launch(activity, uri, customTabsOptions)) {
        return
      }

      val customTabsIntent = customTabsIntentFactory.createIntent(
        activity,
        requireNotNull(customTabsOptions),
        customTabsSessionManager
      )
      if (partialCustomTabsLauncher.launch(activity, uri, customTabsIntent)) {
        return
      }
      customTabsIntent.launchUrl(activity, uri)
    } catch (e: ActivityNotFoundException) {
      throw FlutterError(CODE_LAUNCH_ERROR, e.message, null)
    }
  }

  override fun closeAllIfPossible() {
    val activity = this.activity ?: return
    if (Build.VERSION.SDK_INT < Build.VERSION_CODES.M) {
      return
    }

    val am = activity.getSystemService<ActivityManager>()
    val selfActivityName = ComponentName(activity, activity.javaClass)
    for (appTask in requireNotNull(am).appTasks) {
      val taskInfo = appTask.taskInfo
      if (selfActivityName != taskInfo.baseActivity || taskInfo.topActivity == null) {
        continue
      }

      val serviceIntent = Intent(ACTION_CUSTOM_TABS_CONNECTION)
        .setPackage(taskInfo.topActivity?.packageName)
      if (resolveService(activity.packageManager, serviceIntent) != null) {
        try {
          val intent = Intent(activity, activity.javaClass)
            .setFlags(FLAG_ACTIVITY_CLEAR_TOP or FLAG_ACTIVITY_SINGLE_TOP)
          activity.startActivity(intent)
        } catch (ignored: ActivityNotFoundException) {
        }
        break
      }
    }
  }

  override fun warmup(options: Map<String, Any>?): String? {
    val activity = this.activity ?: return null

    val sessionOptions = customTabsSessionManager.createSessionOptions(options)
    val sessionController =
      customTabsSessionManager.createSessionController(activity, sessionOptions)
        ?: return null

    return if (sessionController.bindCustomTabsService(activity)) {
      sessionController.packageName
    } else {
      null
    }
  }

  override fun mayLaunch(urls: List<String>, sessionPackageName: String) {
    val controller = customTabsSessionManager.getSessionController(sessionPackageName) ?: return
    controller.mayLaunchUrls(urls)
  }

  override fun invalidate(sessionPackageName: String) {
    customTabsSessionManager.invalidateSession(sessionPackageName)
  }

  /**
   * @noinspection SameParameterValue
   */
  @Suppress("deprecation")
  private fun resolveService(
    pm: PackageManager,
    intent: Intent,
    flags: Int = 0
  ): ResolveInfo? {
    return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
      pm.resolveService(
        intent,
        PackageManager.ResolveInfoFlags.of(flags.toLong())
      )
    } else {
      pm.resolveService(intent, flags)
    }
  }

  private companion object {
    const val CODE_LAUNCH_ERROR = "LAUNCH_ERROR"
  }
}
